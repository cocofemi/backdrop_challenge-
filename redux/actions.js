import * as types from './actionTypes';

export const add_favorite = (data) => dispatch => {
    dispatch({
        type: types.ADD_FAVORITE,
        payload: data
    })
}   

export const remove_favorite = (id) => dispatch => {
    dispatch({
        type: types.REMOVE_FAVORITE,
        payload:{
            id
        }
    })
}