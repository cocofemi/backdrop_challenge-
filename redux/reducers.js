import * as types from './actionTypes'

const initialState = {
    favorites: []
}
export default function favoriteReducer (state = initialState, action) {
    switch(action.type) {
        case types.ADD_FAVORITE: 
            return{...state, favorites:[...state.favorites, action.payload]}
        case types.REMOVE_FAVORITE:
            return{
                ...state,
                favorites: state.favorites.filter(favorite => favorite.id !== action.payload.id)
            }
        default: 
            return state
    }
}