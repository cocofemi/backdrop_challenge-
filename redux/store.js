import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import favoriteReducer from './reducers';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { persistStore, persistReducer } from 'redux-persist';
import {STORAGE_KEY} from "@env";

const persistConfig = {
    key: STORAGE_KEY,
    storage: AsyncStorage,
    whitelist: ['favorites']
}

const rootReducer = combineReducers({
    favorite: persistReducer(persistConfig, favoriteReducer)
})

export const store = createStore(rootReducer, applyMiddleware(thunk));
export const persistor = persistStore(store);