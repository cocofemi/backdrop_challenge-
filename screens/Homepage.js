import React from "react";
import { StatusBar } from 'expo-status-bar';
import { Text, View, StyleSheet, FlatList, Image, TouchableOpacity, Platform} from "react-native";
import {globalStyles} from '../styles/globalStyles';
import { MaterialIcons } from '@expo/vector-icons';
import { EvilIcons } from '@expo/vector-icons';
import { useSelector, useDispatch } from 'react-redux';
import { add_favorite, remove_favorite } from "../redux/actions";
import {widthPercentageToDP as wp} from 'react-native-responsive-screen';
import Spinner from 'react-native-loading-spinner-overlay';
import { fetchCats } from "../api/api";
import {useQuery} from "react-query";

function Homepage() {
    const { data, isLoading, error } = useQuery('cats', fetchCats)

    const { favorites} = useSelector(state => state.favorite)

    const dispatch = useDispatch()
    const handleAddFavorite = async (item) => {
        dispatch(add_favorite(item))
    };

    const handleRemoveFavorite = (id) => {
        dispatch(remove_favorite(id))
    }

    const alreadyFavorited = (id) => {
        if (favorites.filter(favorite => favorite.id === id).length > 0) {
            return true
        }
        return false
    }
    const EmptyListMessage = ({ item }) => {
        return <Text testID="no-results">Sorry, no results found.</Text>
      };

    return (  
        <>
            { isLoading ? (
                <View>
                    <Spinner
                    visible={isLoading}
                    textContent={'Loading...'}
                    textStyle={{ color: '#fff'}}
                    testID="loading-spinner"
                    />
                </View>
            ) : error ? (
                <b>There's an error: {error.message}</b>
            ) : data ? (
                <View style={styles.mainView}>
                    <StatusBar style="auto" />
                    <Text style={{...globalStyles.headerText}}> All Cats</Text>
                    <FlatList 
                        contentContainerStyle={{ paddingBottom: 20 }}
                        initialNumToRender={10}
                        keyExtractor={(item) => item.id}
                        data={data}
                        extraData={data}
                        renderItem={({ item, index }) => (
                            <>
                                <View style={styles.listItemView}>
                                    <Image source={{uri: `${item.image}`}} style={styles.displayImage}/>
                                    <Text style={styles.displayName}>{item.name}</Text>
                                    <View style={styles.favouriteView}>
                                        <TouchableOpacity  
                                            testID={`cat-row-${index}`}
                                            onPress={() => 
                                            alreadyFavorited(item.id) ? handleRemoveFavorite(item.id) : handleAddFavorite(item)}>
                                            {alreadyFavorited(item.id) ? 
                                            <MaterialIcons name="favorite" size={30} color={'red'} style={styles.favouriteIcon}/> :
                                            <EvilIcons name="heart" size={30} color="black" style={styles.favouriteIcon}/> }
                                        </TouchableOpacity>
                                    </View>
                                </View>
                            </>
                        )}
                        ListEmptyComponent={EmptyListMessage}
                    />
                </View>
            ) : null }
        </>
    );
}
const styles = StyleSheet.create({
    mainView: {
        width: wp('98%'),
        flex:1
    },
    listItemView: {
        marginTop: 40,
        flexDirection: 'row',
        justifyContent: 'flex-start',
        // alignContent:'flex-end',
        marginLeft: 20,
        marginBottom: -15
    },
    displayName: {
        fontWeight: '600',
        fontSize: 15,
        marginTop: 20,
        marginLeft: 20,
        width: Platform.OS === "ios" ? 220 : 200
    },
    displayImage: {
        width: 60,
        height: 60,
        borderRadius: 10
    },
    favouriteIcon: {
        marginRight: 0,
        marginTop:15
    }
})
export default Homepage;