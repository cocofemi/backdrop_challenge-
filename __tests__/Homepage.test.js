import React  from 'react';
import { cleanup, render, waitFor, fireEvent } from '@testing-library/react-native';
import fetchMock from "jest-fetch-mock"
import { Provider } from 'react-redux';
import { store } from '../redux/store';
import Homepage from '../screens/Homepage';

fetchMock.enableMocks();
afterEach(cleanup);

describe('Testing Homepage', () => {
  it('should render the homepage correctly', () => {
    const tree = render(<Provider store={store}><Homepage /></Provider>)
    expect(tree).toMatchSnapshot();
  }),

  it('should render a list of cats', async() => {
    fetch.mockResponseOnce(
      JSON.stringify([
        { name: 'bengal', image: 'https://cdn2.thecatapi.com/images/ozEvzdVM-.jpg', id: 'beng' },
        { name: 'Aegean', image: 'https://cdn2.thecatapi.com/images/ozEvzdVM-.jpg', id: 'aege' },
      ])
    )

    const {queryByTestId, getByTestId } = render(<Provider store={store}><Homepage /></Provider>)
    expect(queryByTestId('cat-row-0')).toBeNull();

    await waitFor(() => {
      return queryByTestId('cat-row-0');
    });

    expect(getByTestId('cat-row-0'));
  }),

  it('should render a message if data from API is empty', async () => {
    fetch.mockResponseOnce(JSON.stringify([]));
    const { getByTestId } = render(<Provider store={store}><Homepage /></Provider>)
    
    await waitFor(() => {
      return getByTestId('no-results');
    });

    expect(getByTestId('no-results'));
  });
})