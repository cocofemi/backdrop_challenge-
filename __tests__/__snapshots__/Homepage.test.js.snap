// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Homepage should render the homepage correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "width": 735,
    }
  }
>
  <View>
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      statusBarTranslucent={true}
      supportedOrientations={
        Array [
          "landscape",
          "portrait",
        ]
      }
      transparent={true}
      visible={true}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "bottom": 0,
              "flex": 1,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.25)",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <ActivityIndicator
            color="white"
            size="large"
            style={
              Array [
                Object {
                  "flex": 1,
                },
                Object {},
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "bottom": 0,
                  "flex": 1,
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 20,
                    "fontWeight": "bold",
                    "height": 50,
                    "top": 80,
                  },
                  Object {
                    "color": "#fff",
                  },
                ]
              }
            >
              Loading...
            </Text>
          </View>
        </View>
      </View>
    </Modal>
  </View>
  <Text
    style={
      Object {
        "fontSize": 16,
        "fontWeight": "600",
        "marginLeft": 25,
        "marginTop": 80,
      }
    }
  >
     All Cats
  </Text>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      Object {
        "paddingBottom": 20,
      }
    }
    data={Array []}
    extraData={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <Text
        testID="no-results"
      >
        Sorry, no results found.
      </Text>
    </View>
  </RCTScrollView>
</View>
`;
